### DISCUSSION TITLE
Using Primary Key

### DISCUSSION THREAD

**Discussion by jaysonng** [netreconlab/parse-server-swift#47] [Q&A]
So calling functions inside of the .post function of Vapor doesn't have the check for ParseUser who called the request right? So basically calling a query will need .usePrimaryKey option. 

Is there a way to pass the calling user so that queries made still checks if the user is authorized to make queries?

thanks,

**Comment by cbaker6** [comment]
The examples show how to do this. You need to always pass the options to your functions https://github.com/netreconlab/parse-server-swift/blob/be2c6dd2eb90b4ee9c976ab7a27f6e8f8473f994/Sources/ParseServerSwift/routes.swift#L25-L36

**Reply by jaysonng** [reply]
hmm.  thanks! I missed that.

**Reply by jaysonng** [reply]
I still have to use .usePrimaryKey with parseRequest.hydrateUser() to get my queries to work. I still keep getting unauthorized user as an error without it.

```
 if parseRequest.user != nil {
            parseRequest = try await parseRequest.hydrateUser(options: [.usePrimaryKey], request: req)
        }
```

Do you have any idea what I should look into to check and see why I'm unauthorized when all I'm doing is a basic findall Query on GameScore which has no CLP aside from the basic Public and default data it has?

![Screenshot 2023-07-15 at 1 03 52 AM](https://github.com/netreconlab/parse-server-swift/assets/440445/58b9dfd3-1f9d-4a4d-a0ca-68392cd38483)
![Screenshot 2023-07-15 at 1 03 47 AM](https://github.com/netreconlab/parse-server-swift/assets/440445/c3ea7266-a905-4b12-98bf-0fe0b482fd2f)

**Reply by cbaker6** [reply]
I believe you are having two separate issues. “hydrateUser” is only related to your `_User` table. Parse has certain CLP, ACL, and settings there by default to prevent developers from creating insecure apps. `hydrateUser()` is a `fetch` under the hood using the sessionToken of the user who made the call: https://github.com/netreconlab/Parse-Swift/blob/5f2c5f4404ddc8cbb27dc95897ecdcc7f528ec1e/Sources/ParseSwift/Protocols/ParseHookRequestable.swift#L75-L97

So if you are unauthorized, you are either:
1. Making the call from a user on the client side who isn’t logged in, so they don’t have a sessionToken and therefore cannot fetch themselves from the server
2. Making the call from a user on the client side who has an expired sessionToken
3. Modified your CLP and/or ACL on to be too restrictive, preventing the logged in user from fetching themselves

Since you are saying fetchAll doesn’t work with GameScore as well, I suspect your problem is number 1 above as you probably never had a sessionToken. You can verify this by logging to your console `parseRequest`, if the sessionToken is empty you are making calls from a user who isn’t logged in. IMO most cases you should only be using Cloud Code functions for logged in users and rarely using the primaryKey, but that’s up to your discretion as a developer 

Your GameSecure CLP is also very insecure and I don’t recommend it. You should be restricting at least write and add field access to “Authorized” users only. You probably have a setting on your server that’s protecting your table even though your CLP is unrestrictive, possibly even an ACL

**Reply by cbaker6** [reply]
You should also verify the cloud code I linked about with the default parse server image I supply. If you are using your own server, you need to check your settings. The settings for the server I supply is here: https://github.com/netreconlab/parse-server-swift/blob/be2c6dd2eb90b4ee9c976ab7a27f6e8f8473f994/docker-compose.yml#L19-L40

and determined by https://github.com/netreconlab/parse-server-swift/blob/main/parse/index.js

if the server works with the playground code https://github.com/netreconlab/Parse-Swift/blob/main/ParseSwift.playground/Pages/10%20-%20Cloud%20Code.xcplaygroundpage/Contents.swift, then ParseServerSwift is working correctly and something is wrong with your setup.

**Reply by jaysonng** [reply]
> Your GameSecure CLP is also very insecure and I don’t recommend it. You should be restricting at least write and add field access to “Authorized” users only. You probably have a setting on your server that’s protecting your table even though your CLP is unrestrictive, possibly even an ACL

yes, this was for testing purposes only. I have ACL and CLP all over my Classes.

> Making the call from a user on the client side who isn’t logged in, so they don’t have a sessionToken and therefore cannot fetch themselves from the server

I tried erasing all my entries in the _SESSION class and tried testing again, this time around, the error I received was "invalid session token". So this confirms I was logged in properly before. I re-logged in to test further. 

> You can verify this by logging to your console parseRequest, if the sessionToken is empty you are making calls from a user who isn’t logged in. 

Checked and verified that sessionToken exists and matches with the current one via ParseDashboard. 

as of now, I've boiled it down to somewhere inside  .hydrateUser()  -- maybe user.fetch() that's giving me the unauthorized error. 

I'll have to continue tomorrow.

**Comment by jaysonng** [comment]
> if the server works with the playground code https://github.com/netreconlab/Parse-Swift/blob/main/ParseSwift.playground/Pages/10%20-%20Cloud%20Code.xcplaygroundpage/Contents.swift, then ParseServerSwift is working correctly and something is wrong with your setup.

I'll look into my setup again if I missed any of these, which probably is the problem. My ParseServer setup is super basic and taken from the ReadMe files. I'll go through config soon. 

thanks!
